### PR TITLE
esConsumes migration for EDModules in SiStrip packages, part 1

### PR DESCRIPTION
**Pull request by pieterdavid**
#### PR description:

esConsumes migration https://github.com/cms-sw/cmssw/issues/31061 for modules in packages with SiStrip in the name, outside `DQM/` (which more or less splits the list in two - the rest will be submitted as a separate PR).
I tried to follow the instructions from https://github.com/cms-sw/cmssw/issues/31061, and keep other changes minimal.
I rebased on an IB after https://github.com/cms-sw/cmssw/pull/31756 was merged and did not change the modules migrated in https://github.com/cms-sw/cmssw/pull/31697 (SimTracker/SiStripDigitizer and SiStripFedZeroSuppression), so in principle there should be no conflicts.

#### PR validation:

compiles, and `runTheMatrix.py -l limited --useInput all` passes - I'm not sure that checks all the transitions/updates, hints on validating that are welcome.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
@pieterdavid  it seems this PR creates a huge number of changes in MEs located at AlCaReco and SiStrip folder for several workflows,
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-10-15-2300+9d4fb1/39373/dqm-histo-comparison-summary.html
Is this understood?

**Comment by pieterdavid** [comment]
@jfernan2 no, that's not at all expected - it should be a purely technical change, so no changes, I must have missed something. I'll investigate

**Comment by pieterdavid** [comment]
I think I found my mistake... is there a way to run the comparisons locally on a few workflows to check?

**Comment by perrotta** [comment]
> I think I found my mistake... is there a way to run the comparisons locally on a few workflows to check?

@pieterdavid , you can use the tools cited in the reco twiki
- https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_outputs_FWLite_style
- https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_outputs_DQM_style

**Comment by mmusich** [comment]
@pieterdavid, I think if you run in a vanilla release and with your modifications some workflows, there is also script that the DQM group maintains that can be used: `compareHistograms.py` (for two DQM files) and  `compareDQMOutput.py` (for multiple workflows from `runTheMatrix.py`)

**Comment by jfernan2** [comment]
The scripts @mmusich (thanks!) refers to:
https://github.com/cms-sw/cmssw/tree/master/DQMServices/FileIO/scripts

**Comment by pieterdavid** [comment]
Thanks @perrotta @mmusich and @jfernan2 ! With the last commit the workflows with many differences (`4.22,4.53,136.731,136.793,136.874,140.53,140.56,1000.0`) went back to at most a few differences (in allocated/deallocated memory).

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by jfernan2** [comment]
+1

File: AnalysisAlgos/SiStripClusterInfoProducer/plugins/SiStripProcessedRawDigiProducer.cc [hunk 1]
```diff
@@ -31,16 +29,16 @@ void SiStripProcessedRawDigiProducer::produce(edm::Event& e, const edm::EventSet
   edm::Handle<edm::DetSetVector<SiStripDigi> > inputDigis;
   edm::Handle<edm::DetSetVector<SiStripRawDigi> > inputRawdigis;
 
-  es.get<SiStripGainRcd>().get(gainHandle);
-  subtractorPed->init(es);
-  subtractorCMN->init(es);
+  gainHandle_ = es.getHandle(gainToken_);
```
Comment (perrotta) [inline AnalysisAlgos/SiStripClusterInfoProducer/plugins/SiStripProcessedRawDigiProducer.cc:32]: Why do you also save  these Handle's? According to https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES the data should be accessed directly with the token, while the handle should be reserved for cases in which "the caller want's to check whether the data product exists", which is not the case here.

Comment (pieterdavid) [reply AnalysisAlgos/SiStripClusterInfoProducer/plugins/SiStripProcessedRawDigiProducer.cc]: I missed a few of those indeed, I went through the list again and removed those without a check for the validity of the handle.

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by smorovic** [comment]
+1

File: RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingFactory.cc [hunk 2]
```diff
@@ -12,24 +13,26 @@
 #include "RecoLocalTracker/SiStripZeroSuppression/interface/FastLinearCMNSubtractor.h"
 #include "RecoLocalTracker/SiStripZeroSuppression/interface/TT6CMNSubtractor.h"
 
-std::unique_ptr<SiStripRawProcessingAlgorithms> SiStripRawProcessingFactory::create(const edm::ParameterSet& conf) {
+std::unique_ptr<SiStripRawProcessingAlgorithms> SiStripRawProcessingFactory::create(const edm::ParameterSet& conf,
+                                                                                    edm::ConsumesCollector&& iC) {
   return std::unique_ptr<SiStripRawProcessingAlgorithms>(
-      new SiStripRawProcessingAlgorithms(create_SubtractorPed(conf),
-                                         create_SubtractorCMN(conf),
+      new SiStripRawProcessingAlgorithms(std::move(iC),
```
Comment (perrotta) [inline RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingFactory.cc:19]: Mostly for my own understanding, what is in C++ the effect of repeatedly `std::move` the same rValue reference to the ConsumesCollector? I couldn't find other similar examples.

Comment (makortel) [reply RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingFactory.cc:19]: The first move leaves the moved-from object into an unspecified but valid state (and e.g. with `std::unique_ptr` the moved-from state is specified, it is null pointer). In general moving from the same object is an error.

For `edm::ConsumesCollector` specifically moving doesn't add any value. It is essentially a pointer, and can be (nowadays) copied cheaply. I'd recommend to just pass `edm::ConsumesCollector` by value, unless there is a desire to pass it through some intermediate class with forward declaration only, in which case by (regular/lvalue) reference would be the way to go (unless that function should bind also a temporary `ConsumesCollector`, in which case rvalue reference would be needed).

Comment (perrotta) [reply RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingFactory.cc:19]: @pieterdavid @makortel : then, what is the suggestion/plan for this part of the code?

Comment (pieterdavid) [reply RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingFactory.cc:19]: I followed existing code where I saw it mostly being taken by rvalue reference, but if it doesn't matter and copies are cheap I'll change it to pass it by value or lvalue reference, as @makortel suggested, in this kind of cases.

Comment (pieterdavid) [reply RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingFactory.cc]: This should be fixed with the latest changes (passing the consumescollector by value between the zero-suppression factory methods and constructors); I also made initialising the token optional again in the few cases where it was only used depending on some flag (that should be a bit more robust in the case they are run without the full database, even if I don't think this is done in practice)

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
please test

**Comment by perrotta** [comment]
+1
- It implements the esConsumes migration for EDModules in SiStrip packages 
- Jenkinst tests pass and show no differences in outputs

**Comment by jfernan2** [comment]
+1

**Comment by santocch** [comment]
+1

**Comment by yuanchao** [comment]
+1
* Static tests warnings not from the parts of modifications

**Comment by silviodonato** [comment]
@ggovi @cms-sw/db-l2 do you have any comments?

**Comment by ggovi** [comment]
+1

**Comment by silviodonato** [comment]
@cms-sw/daq-l2 ?

**Comment by smorovic** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**